File: sdks/ios/Sources/RynL10n/IntRange.swift
import Foundation

/// 정수(빌드 넘버) 범위 매칭 — M4. 포함 정수 하한/상한으로 정규화(정확한 인접 판정).
/// semver와 동일한 부분집합 문법. 미지원 문법(`||`·`^`·`~`·범위표기)은 거부.
public enum IntRange {
    public struct Comparator { public let op: String; public let n: Int }
    public struct Interval { public let lower: Int?; public let upper: Int? }

    public enum IntRangeError: Error { case unsupported(String); case invalid(String) }

    public static func parse(_ input: String) throws -> [Comparator] {
        let raw = input.trimmingCharacters(in: .whitespaces)
        if raw.isEmpty { throw IntRangeError.invalid("빈 정수 범위식") }
        for pat in ["||", "^", "~", " - "] where raw.contains(pat) { throw IntRangeError.unsupported(pat) }
        return try raw.split(whereSeparator: { $0 == " " }).map { tok in
            guard let m = firstMatch(String(tok), #"^(>=|<=|>|<|=)?(-?\d+)$"#), let n = Int(m[2]!) else {
                throw IntRangeError.invalid(String(tok))
            }
            return Comparator(op: m[1] ?? "=", n: n)
        }
    }

    public static func interval(_ comparators: [Comparator]) -> Interval {
        var lower: Int?; var upper: Int?
        for c in comparators {
            switch c.op {
            case ">=": if lower == nil || c.n > lower! { lower = c.n }
            case ">": let v = c.n + 1; if lower == nil || v > lower! { lower = v }
            case "<=": if upper == nil || c.n < upper! { upper = c.n }
            case "<": let v = c.n - 1; if upper == nil || v < upper! { upper = v }
            default: if lower == nil || c.n > lower! { lower = c.n }; if upper == nil || c.n < upper! { upper = c.n }
            }
        }
        return Interval(lower: lower, upper: upper)
    }

    public static func satisfies(_ n: Int, _ comparators: [Comparator]) -> Bool {
        let iv = interval(comparators)
        if let lo = iv.lower, n < lo { return false }
        if let hi = iv.upper, n > hi { return false }
        return true
    }

    public static func inRange(_ n: Int, _ range: String) -> Bool {
        (try? satisfies(n, parse(range))) ?? false
    }

    public static func overlaps(_ a: Interval, _ b: Interval) -> Bool {
        let lo: Int? = a.lower == nil ? b.lower : (b.lower == nil ? a.lower : Swift.max(a.lower!, b.lower!))
        let hi: Int? = a.upper == nil ? b.upper : (b.upper == nil ? a.upper : Swift.min(a.upper!, b.upper!))
        guard let lo, let hi else { return true }
        return lo <= hi
    }
}

File: sdks/android/src/main/kotlin/com/rynl10n/IntRange.kt
package com.rynl10n

/** 정수(빌드 넘버) 범위 매칭 — M4. 포함 정수 하한/상한 정규화(정확한 인접 판정). */
object IntRangeMatch {
    data class Comparator(val op: String, val n: Int)
    data class Interval(val lower: Int?, val upper: Int?)

    class IntRangeException(message: String) : Exception(message)

    private val COMPARATOR = Regex("""^(>=|<=|>|<|=)?(-?\d+)$""")

    fun parse(input: String): List<Comparator> {
        val raw = input.trim()
        if (raw.isEmpty()) throw IntRangeException("빈 정수 범위식")
        for (pat in listOf("||", "^", "~", " - ")) if (raw.contains(pat)) throw IntRangeException("미지원 정수 범위 문법 \"$pat\"")
        return raw.split(Regex("\\s+")).map { tok ->
            val m = COMPARATOR.matchEntire(tok) ?: throw IntRangeException("유효하지 않은 정수 비교자 \"$tok\"")
            Comparator(m.groupValues[1].ifEmpty { "=" }, m.groupValues[2].toInt())
        }
    }

    fun interval(comparators: List<Comparator>): Interval {
        var lower: Int? = null
        var upper: Int? = null
        for (c in comparators) when (c.op) {
            ">=" -> if (lower == null || c.n > lower!!) lower = c.n
            ">" -> { val v = c.n + 1; if (lower == null || v > lower!!) lower = v }
            "<=" -> if (upper == null || c.n < upper!!) upper = c.n
            "<" -> { val v = c.n - 1; if (upper == null || v < upper!!) upper = v }
            else -> { if (lower == null || c.n > lower!!) lower = c.n; if (upper == null || c.n < upper!!) upper = c.n }
        }
        return Interval(lower, upper)
    }

    fun satisfies(n: Int, comparators: List<Comparator>): Boolean {
        val iv = interval(comparators)
        if (iv.lower != null && n < iv.lower!!) return false
        if (iv.upper != null && n > iv.upper!!) return false
        return true
    }

    fun inRange(n: Int, range: String): Boolean = satisfies(n, parse(range))

    fun overlaps(a: Interval, b: Interval): Boolean {
        val lo = if (a.lower == null) b.lower else if (b.lower == null) a.lower else maxOf(a.lower, b.lower)
        val hi = if (a.upper == null) b.upper else if (b.upper == null) a.upper else minOf(a.upper, b.upper)
        if (lo == null || hi == null) return true
        return lo <= hi
    }
}
